src/transport/SessionManager.h

```diff
@@ -174,8 +174,155 @@ class DLL_EXPORT SessionManager : public TransportMgrDelegate, public FabricTabl
+            // waiting for a Sigma1 message to arrive). Let's skip those.
+            //
+            if (!session->IsCASESession() || session->GetFabricIndex() == kUndefinedFabricIndex)
+            {
+                return Loop::Continue;
+            }
+
+            auto * compareFabric = mFabricTable->FindFabricWithIndex(session->GetFabricIndex());
+            VerifyOrDie(compareFabric != nullptr);
+
+            err = compareFabric->FetchRootPubkey(comparePubKey);
+            VerifyOrDie(err == CHIP_NO_ERROR);
+
+            if (comparePubKey.Matches(targetPubKey) && targetFabric->GetFabricId() == compareFabric->GetFabricId() &&
+                session->GetPeerNodeId() == node.GetNodeId())
+            {
+                function(session);
+            }
+
+            return Loop::Continue;
+        });
+
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * @brief
+     *
+     * Call the provided lambda on all sessions that match the logical fabric
+     * associated with the provided fabric index.
+     *
+     */
+    template <typename Function>
+    CHIP_ERROR ForEachCollidingSession(FabricIndex fabricIndex, Function && function)
+    {
+        Crypto::P256PublicKey targetPubKey;
+
+        auto * targetFabric = mFabricTable->FindFabricWithIndex(fabricIndex);
+        VerifyOrReturnError(targetFabric != nullptr, CHIP_ERROR_INVALID_FABRIC_INDEX);
+
+        auto err = targetFabric->FetchRootPubkey(targetPubKey);
+        VerifyOrDie(err == CHIP_NO_ERROR);
+
+        mSecureSessions.ForEachSession([&](auto * session) {
+            Crypto::P256PublicKey comparePubKey;
+
+            //
+            // It's entirely possible to either come across a PASE session OR, a CASE session
+            // that has yet to be activated (i.e a CASEServer holding onto a SecureSession object
+            // waiting for a Sigma1 message to arrive). Let's skip those.
+            //
+            if (!session->IsCASESession() || session->GetFabricIndex() == kUndefinedFabricIndex)
+            {
+                return Loop::Continue;
+            }
+
+            auto * compareFabric = mFabricTable->FindFabricWithIndex(session->GetFabricIndex());
+            VerifyOrDie(compareFabric != nullptr);
+
+            err = compareFabric->FetchRootPubkey(comparePubKey);
+            VerifyOrDie(err == CHIP_NO_ERROR);
+
+            if (comparePubKey.Matches(targetPubKey) && targetFabric->GetFabricId() == compareFabric->GetFabricId())
+            {
+                function(session);
+            }
+
+            return Loop::Continue;
+        });
+
+        return CHIP_NO_ERROR;
+    }
+
     void ExpireAllSessions(const ScopedNodeId & node);
     void ExpireAllSessionsForFabric(FabricIndex fabricIndex);
+
+    CHIP_ERROR ExpireAllCollidingSessions(const ScopedNodeId & node);
+    CHIP_ERROR ExpireAllCollidingSessionsForFabric(FabricIndex fabricIndex);
```

Comment:
The original `ExpireAllSessionsForFabric` no longer does what its name implies.

I understand why it is functionally OK to do what we're doing here: the 1:1 mapping between fabric index and fabric identity is only necessary to fulfill the external data contract if serving the Operational Credentials cluster.  However, this is confusion inducing.

If the original methods no longer do what their names imply, I think the names should change.  And if the names arent' changed, perhaps the original methods should have the internal magic to actually `ExpireAllSessionsForFabric` by examining all fabrics with matching identities and expiring sessions for these as well.

Alternatively, if no such additions are made, I don't think the new interfaces should be compiled into the code by default, as the new methods and functionality change the behavior of the existing methods.

Reply:
Discussed offline.  What this PR is running into is that `ExpireAllSessionsForFabric` hasn't changed behavior, but was already a bit of a misnomer, in that it removes all sessions for the given fabric if and only if there is only one identity for the given fabric.  What it is really doing (and this is true both with and without this PR) is removing all sessions associated with the given fabric index.

I would be OK with any reasonable solution, including keeping the code in this PR unchanged, but just making sure there are comments to make behavior clear.

Reply:
I've added more comments to the added functions.